### PR TITLE
Docker Registry is now a Clouddriver provider.

### PR DESCRIPTION
**Pull request by lwander**
The goal is to supply other providers with a way to retrieve cached image and repository details. The next step is to actually cache the tags associated with the repositories a user specifies. This setup works for any docker registry as long as it conforms to the v2 API. Authentication is done using the token passing
scheme described [here](https://docs.docker.com/registry/spec/auth/token/).

The configuration looks like this:
```
dockerRegistry:
  enabled: true
  accounts:
    - name: my-docker-account
      username: lwander
      password: hunter2
      email: lwander@google.com
      address: https://index.docker.io
      repositories:
        - library/ubuntu
        - lwander/private-repository
```

Note, I gutted `kato-docker` as talked about last week.

@duftler, @skim1420, @tomaslin, @cfieber

### CODE REVIEW

**Comment by cfieber** [comment]
what are the chances we both have the same password! crazy world

**Comment by cfieber** [comment]
I haven't done a line-wise review of the code yet (at a speed-read nothing huge jumps out) but the approach seems to make sense to me.

Could you talk about what comes next if this goes in:
* Is there a registry caching agent?
* are there registry operations?
* Using the k8s provider as an example, how does it wire in and use this provider?

**Comment by lwander** [comment]
It doesn't show up as *'s?

1. No caching agent yet, that's next.
2. The only operation so far is `getTags(repository)` which lists all tags associated with a Docker repository. There is also a version check to ensure we are pointing at a Docker v2 repository.
3. The Kubernetes provider will define a list of Docker account names, ie `my-dockerhub-account`, `my-private-docker-registry`, etc, and those will be passed to Deck. At this point, Deck can query the Clouddriver caches maintained by the Docker provider for each Docker account and populate the list of images & tags.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerToken.groovy [hunk 1]
```diff
@@ -14,11 +14,15 @@
...
 
-import com.netflix.spinnaker.kato.docker.model.Image
-import com.netflix.spinnaker.kato.docker.security.Docker
+import groovy.transform.ToString
 
-interface RegistryService {
-  Image getImage(Docker docker, String application, String version)
+@ToString(includeNames = true)
+class DockerBearerToken {
+  // One of token or bearer_token will be filled by the request
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerToken.groovy]: Missing trailing period.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        }
+      } else if (key.equalsIgnoreCase(serviceKey)) {
+        result.service = value
+      } else if (key.equalsIgnoreCase(scopeKey)) {
+        result.scope = value
+      }
+    }
+
+    if (!result.realm) {
+      throw new DockerRegistryAuthenticationException('Www-Authenticate header must provide "realm" parameter.')
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy]: The exception message on line 54 uses single quotes, but you are using double quotes for these 3 exception messages.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClient.groovy [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    @Headers([
+      "User-Agent: Spinnaker-Clouddriver",
+      "Docker-Distribution-API-Version: registry/2.0"
+    ]) // TODO(lwander) get clouddriver version #
+    Response checkVersion()
+  }
+
+  /*
+   * Implements token request flow described here https://docs.docker.com/registry/spec/auth/token/
+   * The tokenService also caches tokens for us, so it will attempt to use an old token before retrying
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClient.groovy]: Missing trailing period.

File: clouddriver-web/config/clouddriver.yml [hunk 4]
```diff
@@ -31,6 +31,7 @@ swagger:
     - .*azure.*
     - .*gce.*
     - .*kubernetes.*
+    - .*dockerRegistry.*
```
Comment (duftler) [inline clouddriver-web/config/clouddriver.yml]: If there are no new controller endpoints exposed, probably don't need this (yet).

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/Main.groovy [hunk 5]
```diff
@@ -24,6 +24,7 @@ import com.netflix.spinnaker.clouddriver.core.RetrofitConfig
 import com.netflix.spinnaker.clouddriver.deploy.config.DeployConfiguration
 import com.netflix.spinnaker.clouddriver.google.GoogleConfiguration
 import com.netflix.spinnaker.clouddriver.kubernetes.KubernetesConfiguration
+import com.netflix.spinnaker.clouddriver.docker.registry.DockerRegistryConfiguration
```
Comment (duftler) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/Main.groovy]: Sort-order is off here.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 6]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2015 Google, Inc.
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy]: s/2015/2016

File: clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClientSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  ObjectMapper mapper = new ObjectMapper()
+
+  @Shared
+  DockerRegistryClient client
+
+  def setupSpec() {
+    client = new DockerRegistryClient("https://index.docker.io", "", "", "")
+  }
+
+  def setup() {
```
Comment (duftler) [inline clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClientSpec.groovy]: Is this necessary?

File: clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClientSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import spock.lang.Shared
+import spock.lang.Specification
+
+class DockerRegistryClientSpec extends Specification {
+  private static final REPOSITORY1 = "library/ubuntu"
+
+  @Shared
+  ObjectMapper mapper = new ObjectMapper()
```
Comment (duftler) [inline clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClientSpec.groovy]: Seems extraneous.

File: clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenServiceSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+class DockerBearerTokenServiceSpec extends Specification {
+  private static final REALM1 = "https://auth.docker.io"
+  private static final PATH1 = "token"
+  private static final SERVICE1 = "registry.docker.io"
+  private static final SCOPE1 = "repository:library/ubuntu:push,pull"
+  private static final SCOPE2 = "repository:library/ubuntu:push"
+  private static final REPOSITORY1 = "library/ubuntu"
+
+  @Shared
+  ObjectMapper mapper = new ObjectMapper()
```
Comment (duftler) [inline clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenServiceSpec.groovy]: Seems extraneous.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    cachedTokens[repository] = token
+    return token
+  }
+
+  private interface TokenService {
+    @GET("/{path}")
+    @Headers([
+      "User-Agent: Spinnaker-Clouddriver",
+      "Docker-Distribution-API-Version: registry/2.0"
+    ]) // TODO(lwander) get clouddriver version #
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy]: You probably want something like this: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/GoogleConfiguration.groovy#L65-L68 and https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/GoogleConfiguration.groovy#L45 and https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/security/GoogleCredentialsInitializer.groovy#L39-L40

Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy]: Can probably do this in a followon PR (and promote the clouddriver version stuff out of the google package).

**Comment by duftler** [comment]
LGTM after addressing comments. Very nice.
Needs a rebase.
@cfieber Any concerns on merging this?

**Comment by cfieber** [comment]
Nope - LGTM as well, thanks @lwander !

**Comment by lwander** [comment]
Squashed & addressed @duftler's comments (minus the version number, I'll do that in a separate PR).

**Pull request merged**